drivers/openxr/openxr_device.h

```diff
@@ -0,0 +1,299 @@
+/*************************************************************************/
+/*  openxr_device.h                                                      */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef OPENXR_DRIVER_H
+#define OPENXR_DRIVER_H
+
+#ifdef OPENXR_DUMMY
+// if OpenXR is not supported on the platform, we still have other parts of Godot
+// attempting to use the OpenXR driver, so we need a dummy
+
+class OpenXRDevice {
+public:
+	static void setup_global_defs(){};
+	static bool openxr_is_enabled() { return false; };
+	static OpenXRDevice *get_singleton() { return nullptr; };
+
+	bool is_initialized() { return false; };
+	bool is_running() { return false; };
+	bool initialise(const String &p_rendering_driver) { return false; };
+	bool initialise_session() { return false; };
+	void finish(){};
+};
+
+#else
```

Comment:
How complex would it be to turn `OpenXRDevice` into an interface and have a `dummy` and a `concrete` implementation similar to the pattern used in other areas?
That'll ensure the dummy version stays up-to-date as this class grows in functionality.

Reply:
My main reason for not doing this is that the implementation is purely based on OpenXR, there really isn't a way to abstract things to some sort of template class that makes a lot of sense and especially for the vulkan extension, we'd be implementing far more on the dummy class while it is never even instantiated (the whole point is that if OpenXR is not available or not running, there will never be an instance).

This really is uniquely OpenXR and frankly frankenstein enough, while I understand why OpenXR went down this path, I really don't think they've architectures this very nicely. From a principled view they way OpenXR sets its hooks into Vulkan only makes sense for a purpose build XR engine.